src/Dictionaries/CacheDictionary.cpp

```diff
@@ -339,33 +403,35 @@ void CacheDictionary::has(const PaddedPODArray<Key> & ids, PaddedPODArray<UInt8>
         for (const auto row : ext::range(0, rows))
         {
             const auto id = ids[row];
-            const auto find_result = findCellIdx(id, now);
-            const auto & cell_idx = find_result.cell_idx;
+            const auto find_result = findCellIdxForGet(id, now);
+            auto & cell = cells[find_result.cell_idx];
 
             auto insert_to_answer_routine = [&] ()
             {
-                out[row] = !cells[cell_idx].isDefault();
+                out[row] = !cell.isDefault();
             };
 
             if (!find_result.valid)
             {
                 if (find_result.outdated)
                 {
                     /// Protection of reading very expired keys.
-                    if (now > cells[find_result.cell_idx].strict_max)
+                    if (isExpiredPermanently(now, cell.expiresAt()))
```

Comment:
What is 'permanently expired'? An expired value that is so old we don't read it even when expired reads are allowed? It would be good to describe somewhere all of the states a cell can have.

Reply:
Removed it and added a state of result - not found, found, expired, expired_permanently.